src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionScopeInternal.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ClientConnectionScopeProperties
+    {
+        public IServiceConnection ServiceConnection { get; set; }
+
+        // todo: add additional properties to flow with the scope here
+    }
+
+    internal class ServiceConnectionScopeHolder
+    {
+        private static AsyncLocal<ServiceConnectionScopeHolder> s_AsyncLocalSCC = new AsyncLocal<ServiceConnectionScopeHolder>();
+        internal static ServiceConnectionScopeHolder Current
+        {
+            get => s_AsyncLocalSCC.Value;
+            set => s_AsyncLocalSCC.Value = value;
+        }
+
+        public ClientConnectionScopeProperties Properties { get; set; }
+    }
+
+    internal class ServiceConnectionScopeInternal : IDisposable
+    {
+        readonly private bool _needCleanup = false;
+
+        public ServiceConnectionScopeInternal() : this(default)
+        {
+        }
+
+        internal ServiceConnectionScopeInternal(ClientConnectionScopeProperties properties)
+        {
+            if (ServiceConnectionScopeHolder.Current == null)
+            {
+                ServiceConnectionScopeHolder.Current = new ServiceConnectionScopeHolder() { Properties = properties };
+                _needCleanup = true;
+            }
+            else
+            {
+                Debug.Assert(properties == null,
+                    $"Attempt to replace existing connection scope  {ServiceConnectionScopeHolder.Current.Properties.ServiceConnection?.GetHashCode()} with new connection: {properties?.ServiceConnection?.GetHashCode()}");
+            }
+        }
+
+        internal static ServiceConnectionScopeHolder Holder => ServiceConnectionScopeHolder.Current;
```

Comment:
Is there a need to expose this `Holder` for the scope? I feel `Holder` as some internal implementation of the scope that we can hide from the users

Reply:
This is internal class and deep at connection level we will need to know if the client connection scope has been established (so when we need to choose or replace the connection we know that it will persist). See https://github.com/Azure/azure-signalr/pull/658/files#diff-72c9ae4e6a93ede091bc3b08bae82afaR299 in the old PR.

The additional user-facing class `ServiceCommunicationScope<T>` (or its sub-classes) does not expose the holder anymore.